### PR TITLE
Enhancement: Uplift to TS26.512 v17.7.0

### LINKED ISSUE #112: M1: Uplift to TS 26.512 V17.7.0
# Background

The following changes to TS 26.512 were agreed at SA4#126 (Chicago) and approved at SA#102 (Edinburgh):

* CR0054 "Essential maintenance" in [S4-231631](https://www.3gpp.org/ftp/TSG_SA/WG4_CODEC/TSGS4_126_Chicago/Docs/S4-231631.zip) and plenary CR Pack [SP-231368](https://www.3gpp.org/ftp/TSG_SA/TSG_SA/TSGS_102_Edinburgh_2023-12/Docs/SP-231368.zip).

The updated specification will be published shortly as TS 26.512 V17.7.0.

This issue captures the changes that need to be made to the reference implementation of the 5GMS AF Provisioning API at reference point M1.

# Summary of changes

1.  _applicationSessionContext_ becomes an optional property of the Policy Template resource in **Policy Templates Provisioning API**.
2. **Policy Templates Provisioning API** now has `readOnly: true`  for _policyTemplateId_, _state_ and _stateReason_.
3. HTTP response code changed for malformed regular expression error in cache purge operation of **Content Hosting Provisioning API** (clauses 7.6.4.3 and C.3.5).
    - *422 (Unprocessable Content)* becomes *400 (Bad Request)*.
4. **Consumption Reporting Provisioning API** enforces boundaries for _reportingInterval_.
    -  `minimum: 0`.
    -  `exclusiveMinimum: true`.

Until the the OpenAPI YAML is updated, the changes are available on the draft branch:

- https://forge.3gpp.org/rep/all/5G_APIs/-/tree/Rel17-draft-TSG102

### LINKED ISSUE #113: M5 - Service Access Information: Uplift to TS 26.512 V17.7.0
# Background

The following changes to TS 26.512 were agreed at SA4#126 (Chicago) and approved at SA#102 (Edinburgh):

* CR0054 "Essential maintenance" in [S4-231631](https://www.3gpp.org/ftp/TSG_SA/WG4_CODEC/TSGS4_126_Chicago/Docs/S4-231631.zip) and plenary CR Pack [SP-231368](https://www.3gpp.org/ftp/TSG_SA/TSG_SA/TSGS_102_Edinburgh_2023-12/Docs/SP-231368.zip).

The updated specification will be published shortly as TS 26.512 V17.7.0.

This issue captures the changes that need to be made to the reference implementation of the 5GMS AF Service Access Information API at reference point M5.

# Summary of changes

1. Due to the change in the data type, the AF now needs to de-reference the reporting interval for the Consumption Reporting Configuration associated with the Provisioning Session.
2. The AF now needs to check if `is_access_reporting` is present in the M1 Consumption Reporting Configuration and should provide the specified default value at M5 if that Boolean is false.
3. The AF now needs to check if `is_access_reporting` is present in the M1 Consumption Reporting Configuration and should provide the specified default value at M5 if that Boolean is false (The Media Session Handler reference implementation needs to be changed as well).

Until the the OpenAPI YAML is updated, the changes are available on the draft branch:

- https://forge.3gpp.org/rep/all/5G_APIs/-/tree/Rel17-draft-TSG102

### LINKED ISSUE #115: M5 Consumption Reporting - ERROR: Failed to store Consumption Report for provisioning session
## Description
I am currently testing the end-to-end chain for consumption reporting . This includes:

* Version 1.1.0 of the UE components (currently in the `development` branch)
* Version 1.4.0 of the AF (currently in the `development` branch)
* Version 1.2.2 of the AS (currently in the `development` branch)

As part of the testing I identified a minor issue in the Media Session Handler as documented here: https://github.com/5G-MAG/rt-5gms-media-session-handler/issues/52

After fixing the issue above the consumption reports are accepted by the AF and I can see a single report being created:

````json
{
    "mediaPlayerEntry": "http://192.168.2.7/m4d/provisioning-session-527136fe-b07f-41ee-b0e8-039343ecf354/bbb/2/client_manifest-common_init.mpd",
    "reportingClientId": "5edfc1ad-40ac-4b1e-8380-bc5fd466e223",
    "consumptionReportingUnits":
    [
        {
            "mediaConsumed": "1920x1080p25",
            "mediaEndpointAddress":
            {
                "ipv4Addr": "192.168.2.4",
                "portNumber": 80
            },
            "startTime": "2024-01-11T13:46:37Z",
            "duration": 9,
            "locations":
            []
        },
        {
            "mediaConsumed": "160kbps",
            "mediaEndpointAddress":
            {
                "ipv4Addr": "192.168.2.4",
                "portNumber": 80
            },
            "startTime": "2024-01-11T13:46:37Z",
            "duration": 9,
            "locations":
            []
        }
    ]
}
````

After creating the first consumption report the Application Function is throwing an error for each of the following reports:

````
01/11 12:48:07.400: [msaf] ERROR: Failed to store Consumption Report for provisioning session [527136fe-b07f-41ee-b0e8-039343ecf354] (../src/5gmsaf/msaf-m5-sm.c:793)
01/11 12:48:18.264: [msaf] ERROR: Unable to create /var/local/log/open5gs/reports/527136fe-b07f-41ee-b0e8-039343ecf354/consumption_reports/5edfc1ad-40ac-4b1e-8380-bc5fd466e223_2024-01-11T13:46:37Z.json for writing: File exists (../src/5gmsaf/data-collection.c:111)
01/11 12:48:18.264: [msaf] ERROR: Failed to store Consumption Report for provisioning session [527136fe-b07f-41ee-b0e8-039343ecf354] (../src/5gmsaf/msaf-m5-sm.c:793)
01/11 12:48:27.492: [msaf] ERROR: Unable to create /var/local/log/open5gs/reports/527136fe-b07f-41ee-b0e8-039343ecf354/consumption_reports/5edfc1ad-40ac-4b1e-8380-bc5fd466e223_2024-01-11T13:46:37Z.json for writing: File exists (../src/5gmsaf/data-
````

It looks like for some reason the time in the filename is not increased and the existing file is not overwritten. The response code of the Application Function is 500.

### PR DESCRIPTION
**Pull request by davidjwbbc**
This updates the code in line with changes agreed for 3GPP TS 26.512 v17.7.0.

Closes #112 
Closes #113
Closes #115 

Please see #112 and #113 for full details of the v17.7.0 changes, and #115 for the changes needed to accommodate the clarification of consumption reporting in v17.7.0.

### CODE REVIEW

**Review by devbbc — APPROVED** [review]

**Review by rjb1000 — APPROVED** [review]

**Comment by davidjwbbc** [comment]
Changes in the recent specs adjust the understanding of consumption reporting, so a fix for #115 will also be included in this uplift.

**Comment by davidjwbbc** [comment]
Added @dsilhavy  as a reviewer as this covers consumption reporting storage fixes and highlights changes to M5 interfaces.

File: tools/python3/lib/rt_m1_client/types.py [hunk 1]
```diff
@@ -46,11 +46,11 @@ def wrapped_default(self, obj):
 ProvisioningSessionType = Literal['DOWNLINK','UPLINK']
 
 class ProvisioningSessionMandatory(TypedDict):
-    '''Madatory fields for a `ProvisioningSession`
+    '''Madatory fields for a `ProvisioningSession` v17.7.0
```
Comment (rjb1000) [inline tools/python3/lib/rt_m1_client/types.py:49]: Typo "Madatory".

File: src/5gmsaf/msaf-m5-sm.c [hunk 2]
```diff
@@ -776,11 +776,15 @@ void msaf_m5_state_functional(ogs_fsm_t *s, msaf_event_t *e)
 
                                             consumption_report = msaf_api_consumption_report_parseRequestFromJSON(json, &reason);
                                             if (consumption_report) {
+                                                struct timespec ts;
+                                                char buf[32];
+                                                char *filetime = NULL;
+                                                clock_gettime(CLOCK_REALTIME, &ts);
+                                                strftime(buf, sizeof(buf), "%Y-%m-%dT%H:%M:%S", gmtime(&ts.tv_sec));
+                                                filetime = ogs_msprintf("%s.%.6iZ", buf, (int)(ts.tv_nsec/1000));
```
Comment (rjb1000) [inline src/5gmsaf/msaf-m5-sm.c:784]: Looks good. Thanks, @davidjwbbc.

**Review by rjb1000 — APPROVED** [review]

File: tools/python3/lib/rt_m1_client/types.py (hunk 1 above)
Comment (davidjwbbc) [reply tools/python3/lib/rt_m1_client/types.py]: That's been in there a while! Fixed!

**Review by dsilhavy — APPROVED** [review]
Confirmed that this fixes #115

**Pull request merged**